src/adventure/adventuretracker.cpp

```diff
@@ -0,0 +1,197 @@
+        auto achievement = m_achievementParser.getLastSuccessVal();
+        auto xpGained = checkpointXP();
+        emit sig_achievedSomething(achievement, xpGained);
+        return;
+    }
+
+    if (m_accomplishedTaskParser.parse(line)) {
+        auto xpGained = checkpointXP();
+        emit sig_accomplishedTask(xpGained);
+        return;
+    }
+
+    if (m_diedParser.parse(line)) {
+        auto xpLost = checkpointXP();
+        emit sig_diedInGame(xpLost);
+        return;
+    }
+
+    if (m_hintParser.parse(line)) {
+        auto hint = m_hintParser.getLastSuccessVal();
+        emit sig_receivedHint(hint);
+        return;
+    }
+}
+
+void AdventureTracker::slot_onUserGmcp(const GmcpMessage &msg)
+{
+    // https://mume.org/help/generic_mud_communication_protocol
+
+    if (msg.isCharName()) {
+        parseIfUpdatedCharName(msg);
+        return;
+    }
+
+    if (msg.isCharStatusVars()) {
+        parseIfUpdatedCharName(msg);
+        return;
+    }
+
+    if (msg.isCharVitals()) {
+        parseIfUpdatedVitals(msg);
+        return;
+    }
+
+    if (msg.isCommChannelText()) {
+        parseIfReceivedComm(msg);
+        return;
+    }
+
+    if (msg.isCoreGoodbye()) {
+        parseIfGoodbye(msg);
+        return;
+    }
+}
+
+void AdventureTracker::parseIfGoodbye([[maybe_unused]] GmcpMessage msg)
+{
+    if (!m_session)
+        return;
+
+    qDebug().noquote() << QString("Adventure: ending session for %1").arg(m_session->name());
+    m_session->endSession();
+    emit sig_endedSession(*m_session);
+    m_session.reset();
+}
+
+void AdventureTracker::parseIfReceivedComm(GmcpMessage msg)
+{
+    std::optional<QJsonDocument> doc = msg.getJsonDocument();
+    if (!doc || !doc->isObject())
+        return;
+    QJsonObject obj = doc->object();
+
+    if (!obj.contains("channel") || !obj.contains("text"))
+        return;
+
+    if (obj["channel"].toString() == "tells") {
+        emit sig_receivedTell(obj["text"].toString());
+
+    } else if (obj["channel"].toString() == "tales") {
+        emit sig_receivedNarrate(obj["text"].toString());
+    }
+}
+
+void AdventureTracker::parseIfUpdatedCharName(GmcpMessage msg)
+{
+    std::optional<QJsonDocument> doc = msg.getJsonDocument();
+    if (!doc || !doc->isObject())
+        return;
+    QJsonObject obj = doc->object();
+
+    if (!obj.contains("name")) {
+        return;
+    }
+
+    auto charName = obj["name"].toString();
+
+    if (!m_session) {
+        qDebug().noquote() << QString("Adventure: new adventure for %1").arg(charName);
+
+        m_session = std::make_unique<AdventureSession>(charName);
+        emit sig_updatedSession(*m_session);
+        return;
+    }
+
+    if (m_session && m_session->name() != charName) {
+        qDebug().noquote() << QString("Adventure: new adventure for %1 replacing %2")
+                                  .arg(charName, m_session->name());
+
+        m_session->endSession();
+        emit sig_endedSession(*m_session);
```

Comment:
This feels scary because the old `m_session` is destructed after this and whomever is listening might try to keep it around. I wonder if a `std::shared_ptr` would be safer.